### PR TITLE
Fix cctor abort race

### PR DESCRIPTION
**Pull request by kumpera**

### CODE REVIEW

File: mono/metadata/object-internals.h [hunk 1]
```diff
@@ -369,7 +369,7 @@ struct _MonoInternalThread {
 	int _serialized_principal_version;
 	gpointer appdomain_refs;
 	/* This is modified using atomic ops, so keep it a gint32 */
-	gint32 interruption_requested;
+	gint32 __interruption_requested;
 	MonoCoopMutex *synch_cs;
```
Comment (vargaz) [inline mono/metadata/object-internals.h]: Why does this start with an underscore ?

File: mono/metadata/object.c [hunk 2]
```diff
@@ -380,6 +382,18 @@ mono_runtime_class_init_full (MonoVTable *vtable, MonoError *error)
...
 		return FALSE;
 	}
+	if (!mono_threads_try_begin_abort_protected_block ()) {
+		mono_type_initialization_unlock ();
+		MonoException *exc = mono_thread_interruption_checkpoint ();
+		if (!exc) {
+			if (retry_count > 10)
+				g_error ("Retry count is too big (%d), this should not happen", retry_count);
+			goto retry_init;
+		}
```
Comment (vargaz) [inline mono/metadata/object.c]: What does retrying solve ?

File: mono/metadata/threads.c [hunk 3]
```diff
@@ -756,6 +783,7 @@ static guint32 WINAPI start_wrapper_internal(StartInfo *start_info, gsize *stack
 		void *args [1];
 
 		g_assert (start_delegate != NULL);
+		mono_thread_info_sleep (3, NULL);
```
Comment (vargaz) [inline mono/metadata/threads.c]: Leftover.

File: mono/metadata/threads.c [hunk 4]
```diff
@@ -5182,6 +5210,20 @@ mono_threads_begin_abort_protected_block (void)
 	mono_memory_barrier ();
 }
 
+
+/* Returns  FALSE if there's a pending interruption */
```
Comment (vargaz) [inline mono/metadata/threads.c]: Extra whitespace line.

File: mono/metadata/threads.c [hunk 5]
```diff
@@ -5182,6 +5210,20 @@ mono_threads_begin_abort_protected_block (void)
...
+mono_threads_try_begin_abort_protected_block (void)
+{
+	MonoInternalThread *thread;
+
+	thread = mono_thread_internal_current ();
+	LOCK_THREAD (thread); //the next call needs it as it accesses MonoInternalThread::state
+	gboolean res = mono_thread_try_inc_abort_prot_block_count (thread);
+	UNLOCK_THREAD (thread);
+	mono_memory_barrier ();
+	return res;
```
Comment (vargaz) [inline mono/metadata/threads.c]: Why is this membar needed ?

Comment (kumpera) [inline mono/metadata/threads.c]: it's a leftover. I'll remove it.

File: mono/metadata/object.c (hunk 2 above)
Comment (kumpera) [inline mono/metadata/object.c]: There are a few problems:

1) we must not execute interrupt while holding the type initialization. As that's a leaf lock.
2) interruption can potentially return NULL... for reasons. Some valid, some just me not being super familiar with the code.

My concern is that the unknown parts of (2) could introduce a retry loop and that's what I'm trying to catch.

File: mono/metadata/object-internals.h (hunk 1 above)
Comment (kumpera) [inline mono/metadata/object-internals.h]: the 2 fields are now dead. I'll add a follow up commit that kills them.

**Comment by kumpera** [comment]
Closing as this was fixed by #4274

**Pull request closed without merging**